experimental/iterators/README.md

```diff
@@ -0,0 +1,146 @@
+end of the sequence, binding each resulting value to the loop variable. This
+roughly corresponds to:
+
+```python3
+iterable = CountToThree()
+iterator = iterable.__iter__()
+print(iterator.__next__()) # Outputs: 1
+print(iterator.__next__()) # Outputs: 2
+print(iterator.__next__()) # Outputs: 3
+print(iterator.__next__()) # Raises: StopIteration
+```
+
+One property that makes iterators in Python so powerful is that they are
+composable. Many
+[built-in functions](https://docs.python.org/3/library/functions.html) (such as
+`enumerate`, `filter`, `map`, etc.) and standard library functions (most notably
+those defined in
+[`itertools`](https://docs.python.org/3/library/itertools.html)) accept
+arguments of type `iterable` and produce again results that are `iterable`.
+I/O is also often done with `iterable` objects (for example, via
+[`IOBase`](https://docs.python.org/3/library/io.html#io.IOBase)).
+
+Consider the following example, which computes the maximum length of a line with
+an even number of characters of all lines in a given text file:
+
+```python3
+max(
+  filter(lambda x: x % 2 == 0,
+         map(lambda x: len(x),
+             open('file.txt'))))
+```
+
+Each of the functions `max`, `filter`, `map`, and `open` either accepts or
+produces an `iterator` object (or both). The result of the three inner-most
+function is an `iterator` object that has two more `iterator` objects
+recursively nested insise of them. That object is given to `max`, which drives
+the computation by calling `__next__()` repeatably on that object, which, in
+turn, causes a cascade of calls to `__next__()` of the nested `iterator`
+objects. Since the state in each `iterator` object and the number of values
+passed between them is constant, the above example essentially works on
+arbitrarily large files.
+
+### Iterators in Database Systems
+
+Database systems (both relational and non-relational) typically express their
+computations as transformations of streams of data in order to reduce data
+movement between slow and fast memory (e.g., between DRAM and CPU cache).
+Transformations are typically made composable through an "iterator" interface:
+Each iterator produces a stream of data that can be consumed "downstream" one
+element at a time, and, in turn, consumes the streams from its zero or more
+"upstream" iterators to do so -- without knowing anything about its upstream and
+downstream iterators. Each iterator may have complex control flow logic, often
+depending on the data that it consumes, and can manage its own state. The
+iterator interface, thus, passes data *and* control: when one iterator asks its
+upstream iterator for the next element, it passes control to that iterator,
+which returns the control together with the next element. By limiting the number
+of in-flight elements to essentially one, this minimizes the overall state and
+thus data movement.
+
+TODO(ingomueller): explain Volcano iterator interface
```

Comment:
Maybe also something about how inlining producer functions can help execution time, as this currently feels like way too many calls everywhere to be very efficient, IMO.

Maybe also something about batch-based iterators, i.e., bundling together say a cache line of data before returning.